plugin/advisor/mysql/advisor_naming_index_convention.go

```diff
@@ -0,0 +1,257 @@
+type NamingIndexConventionAdvisor struct {
+}
+
+// Check checks for index naming convention.
+func (check *NamingIndexConventionAdvisor) Check(ctx advisor.Context, statement string) ([]advisor.Advice, error) {
+	root, errAdvice := parseStatement(statement, ctx.Charset, ctx.Collation)
+	if errAdvice != nil {
+		return errAdvice, nil
+	}
+
+	level, err := advisor.NewStatusBySchemaReviewRuleLevel(ctx.Rule.Level)
+	if err != nil {
+		return nil, err
+	}
+
+	format, templateList, err := api.UnmarshalNamingRulePayloadAsTemplate(ctx.Rule.Type, ctx.Rule.Payload)
+	if err != nil {
+		return nil, err
+	}
+	checker := &namingIndexConventionChecker{
+		level:        level,
+		ruleType:     ctx.Rule.Type,
+		format:       format,
+		templateList: templateList,
+	}
+	for _, stmtNode := range root {
+		(stmtNode).Accept(checker)
+	}
+
+	if len(checker.adviceList) == 0 {
+		checker.adviceList = append(checker.adviceList, advisor.Advice{
+			Status:  advisor.Success,
+			Code:    common.Ok,
+			Title:   "OK",
+			Content: "",
+		})
+	}
+	return checker.adviceList, nil
+}
+
+type namingIndexConventionChecker struct {
+	adviceList   []advisor.Advice
+	level        advisor.Status
+	ruleType     api.SchemaReviewRuleType
+	format       string
+	templateList []string
+}
+
+func (checker *namingIndexConventionChecker) Enter(in ast.Node) (ast.Node, bool) {
+	indexDataList := checker.getIndexMetaDataList(in)
+
+	for _, indexData := range indexDataList {
+		regex, err := getTemplateRegexp(checker.format, checker.templateList, indexData.metaData)
+		if err != nil {
+			checker.adviceList = append(checker.adviceList, advisor.Advice{
+				Status:  checker.level,
+				Code:    common.Internal,
+				Title:   "Internal error for index naming convention rule",
+				Content: fmt.Sprintf("%q meet internal error %q", in.Text(), err.Error()),
+			})
+			continue
+		}
+		if !regex.MatchString(indexData.index) {
+			checker.adviceList = append(checker.adviceList, advisor.Advice{
+				Status:  checker.level,
+				Code:    common.NamingIndexConventionMismatch,
+				Title:   "Mismatch index naming convention",
+				Content: fmt.Sprintf("%q mismatches index naming convention, expect %q but found %q", in.Text(), checker.format, indexData.index),
+			})
+		}
+	}
+
+	if len(checker.adviceList) == 0 {
+		checker.adviceList = append(checker.adviceList, advisor.Advice{
+			Status:  advisor.Success,
+			Code:    common.Ok,
+			Title:   "OK",
+			Content: "",
+		})
+	}
+
+	return in, false
+}
+
+func (checker *namingIndexConventionChecker) Leave(in ast.Node) (ast.Node, bool) {
+	return in, true
+}
+
+type indexMetaData struct {
+	index    string
+	metaData map[string]string
+}
+
+// getIndexMetaData returns the list of index with meta data.
+func (checker *namingIndexConventionChecker) getIndexMetaDataList(in ast.Node) []*indexMetaData {
+	var res []*indexMetaData
+
+	switch node := in.(type) {
+	case *ast.CreateTableStmt:
+		for _, constraint := range node.Constraints {
```

Comment:
I'm not sure if we should also check the keys in `CREATE TABLE tech_book(id INT PRIMARY KEY, name VARCHAR(20))`. For example, we're forcing the pk name as something like `^pk_{{table}}_{{column_list}}`, so forcing the id name as `pk_tech_book_id` in this SQL is some kind of weird.